core/src/main/scala/org/apache/spark/ui/scope/RDDOperationGraph.scala

```diff
@@ -72,6 +72,22 @@ private[ui] class RDDOperationCluster(val id: String, private var _name: String)
   def getCachedNodes: Seq[RDDOperationNode] = {
     _childNodes.filter(_.cached) ++ _childClusters.flatMap(_.getCachedNodes)
   }
+
+  def canEqual(other: Any): Boolean = other.isInstanceOf[RDDOperationCluster]
+
+  override def equals(other: Any): Boolean = other match {
+    case that: RDDOperationCluster =>
+      (that canEqual this) &&
+          _childClusters == that._childClusters &&
+          id == that.id &&
+          _name == that._name
+    case _ => false
+  }
+
+  override def hashCode(): Int = {
+    val state = Seq(_childClusters, id, _name)
+    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
```

Comment:
Yeah, we had a similar discussion at https://github.com/apache/spark/pull/12157 just now. This is a decent way to write it when there are a lot of fields; for 3 fields, I'm neutral about it. The alternative is `31 * (31 * Objects.hashCode(_childClusters) + Objects.hashCode(id)) + Objects.hashCode(_name)`